modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/ampf_connection_validator.rb

```diff
@@ -0,0 +1,112 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Storages
+  module Peripherals
+    module ConnectionValidators
+      module Nextcloud
+        class AmpfConnectionValidator < BaseValidator
+          using ServiceResultRefinements
+
+          private
+
+          def validate
+            register_checks(
+              :userless_access, :group_folder_presence, :files_request, :group_folder_contents
+            )
+
+            userless_access_denied
+            group_folder_not_found
+            files_request_failed_with_unknown_error
+            with_unexpected_content
+          end
+
+          def userless_access_denied
+            if files.result == :unauthorized
+              fail_check(:userless_access, message(:userless_access_denied))
+            else
+              pass_check(:userless_access)
+            end
+          end
+
+          def group_folder_not_found
+            if files.result == :not_found
+              fail_check(:group_folder_presence, message(:group_folder_not_found))
+            else
+              pass_check(:group_folder_presence)
+            end
+          end
+
+          def files_request_failed_with_unknown_error
+            if files.result == :error
+              error "Connection validation failed with unknown error:\n\t" \
+                    "storage: ##{@storage.id} #{@storage.name}\n\t" \
+                    "request: Group folder content\n\t" \
+                    "status: #{files.result}\n\t" \
+                    "response: #{files.error_payload}"
+
+              fail_check(:files_request, message(:unknown_error))
+            else
+              pass_check(:files_request)
+            end
+          end
+
+          def with_unexpected_content
+            unexpected_files = files.result.files.reject { managed_project_folder_ids.include?(it.id) }
+            return pass_check(:group_folder_contents) if unexpected_files.empty?
+
+            log_extraneous_files(unexpected_files)
+            warn_check(:group_folder_contents, message(:unexpected_content))
+          end
+
+          def log_extraneous_files(unexpected_files)
+            file_representation = unexpected_files.map do |file|
+              "Name: #{file.name}, ID: #{file.id}, Location: #{file.location}"
+            end
+
+            warn "Unexpected files/folder found in group folder:\n\t#{file_representation.join("\n\t")}"
+          end
+
+          def auth_strategy = Registry["nextcloud.authentication.userless"].call
+
+          def managed_project_folder_ids
+            @managed_project_folder_ids ||= ProjectStorage.automatic.where(storage: @storage).pluck(:project_folder_id)
```

Comment:
:green_circle: This is used in a loop where we call `include?`. That's a typical use case where I'd recommend storing this collection as a set, which is optimized for inclusion queries (complexity of O(1) instead of O(n)):

```suggestion
            @managed_project_folder_ids ||= ProjectStorage.automatic.where(storage: @storage).pluck(:project_folder_id).to_set
```

Reply:
True that! Will do it.